akka-http-core/src/main/scala/akka/http/engine/server/HttpServerBluePrint.scala

```diff
@@ -240,4 +266,75 @@ private[http] object HttpServerBluePrint {
         case _ ⇒ ctx.fail(error)
       }
   }
+
+  case class WebsocketSetup(
+    flow: Flow[ByteString, ByteString, Any],
+    publisherKey: StreamUtils.ReadableCell[Publisher[FramePart]],
+    subscriberKey: StreamUtils.ReadableCell[Subscriber[FramePart]],
+    switchSource: Source[SwitchToWebsocketToken.type, Any]) extends WebsocketSwitch {
+    @volatile var switchToWebsocketRef: Option[ActorRef] = None
+
+    def switchToWebsocket(handlerFlow: Flow[FramePart, FramePart, Any])(implicit mat: FlowMaterializer): Unit = {
+      // 1. fill processing hole in the websocket pipeline with user-provided handler
+      Source(publisherKey.value)
+        .via(handlerFlow)
+        .to(Sink(subscriberKey.value))
+        .run()
+
+      // FIXME: is there a race condition between 1 and 2?
+
+      // 2. flip the switch
+      switchToWebsocketRef.get ! TokenSourceActor.Trigger
+    }
+  }
+  def websocketPipeline: WebsocketSetup = {
+    val sinkCell = new StreamUtils.OneTimeWriteCell[Publisher[FramePart]]
+    val sourceCell = new StreamUtils.OneTimeWriteCell[Subscriber[FramePart]]
+
+    val sink = StreamUtils.oneTimePublisherSink[FramePart](sinkCell, "frameHandler.in")
+    val source = StreamUtils.oneTimeSubscriberSource[FramePart](sourceCell, "frameHandler.out")
+
+    val flow =
+      Flow[ByteString]
+        .transform[FramePart](() ⇒ new FrameEventParser)
+        .via(StreamUtils.sinkSourceFlow(sink, source))
```

Comment:
This is related to the materialized value access in the graph: if you can manage to thread the materialized value through the graph in a way that I can get the publisher/subscriber pair (the hole) to the parser to be filled at runtime of the flow then this can change (and all the OneTime... stuff can go). I didn't manage to achieve that yet without having to restructure all kind of things of the whole pipeline, so I would like to postpone any such internal optimization to a later point.